man/ranal2.1
.Dd Oct 27, 2010
.Dt RANAL2 1
.Os
.Sh NAME
.Nm ranal2
.Nd radare2 code analysis commandline frontend
.Sh SYNOPSIS
.Nm ranal2
.Op -BhL
.Op -a arch
.Op -b bits
.Op -l inputlen
.Op -o offset
.Ar hexpairs|-
.Sh DESCRIPTION
ranal2 provides a commandline utility to test and perform some code analysis work on a provided hexpair as argument or from stdin if the argument is '-'.
.Pp
.Bl -tag -width Fl
.It Fl a Ar arch
Set a different architecture (x86, ppc, arm, java)
.It Fl b Ar bits
Specify bit size of registers: 8, 16, 32, 64
.It Fl B
Input is binary, \-l is mandatory. Useful for stdin
.It Fl h
Show help message
.It Fl l Ar len
Specify length of input data
.It Fl L
List all supported code analysis plugins
.It Fl o Ar offset
Offset of the opcode to assemble
.El
.Sh SEE ALSO
.Pp
.Xr radare2(1) ,
.Xr rafind2(1) ,
.Xr rahash2(1) ,
.Xr rabin2(1) ,
.Xr radiff2(1) ,
.Xr ragg2(1) ,
.Xr rarun2(1) ,
.Xr rasm2(1) ,
.Xr rax2(1) ,
.Sh AUTHORS
.Pp
pancake <pancake@nopcode.org>,
nibble <nibble@develsec.org>


ChangeLog
-=== RELEASE 0.8 ===-
changeset:   1625:0aaad74ad9fc
bookmark:    master
tag:         tip
user:        pancake
date:        Fri Jul 15 18:13:00 2011 +0200
files:       configure configure.acr libr/core/core.c libr/io/undo.c r2-bindings/python/r_cons.i.experimental
description:
* Bump release 0.8 - Codename thecakeisapie
* Fix uninitialized memory issue in r_block_resize()
  - Thanks vext01 for reporting


changeset:   1624:1315774824fb
parent:      1622:fc2643c36e30
parent:      1623:95f09764122e
user:        pancake
date:        Fri Jul 15 18:07:56 2011 +0200
description:
* Merge heads


changeset:   1623:95f09764122e
parent:      1620:236b81674ab5
user:        earada
date:        Fri Jul 15 14:17:36 2011 +0200
files:       libr/bin/p/bin_elf.c
description:
* Fix virtual address in elf without setcions


changeset:   1622:fc2643c36e30
user:        pancake
date:        Fri Jul 15 12:57:30 2011 +0200
files:       mk/gitpush.sh
description:
* Fix some typos in gitpush script


changeset:   1621:b6a6d3012998
user:        pancake
date:        Fri Jul 15 12:48:46 2011 +0200
files:       Makefile mk/gitpush.sh
description:
* Added gitpush utility for github mirroring


changeset:   1620:236b81674ab5
user:        pancake@pair
date:        Fri Jul 15 02:21:40 2011 +0200
files:       configure configure.acr
description:
* Bump acr to 0.8.6 with shellscript endian check
  - New flags --with-{little|big}-endian


changeset:   1619:d49dd6f2e789
user:        pancake@pair
date:        Fri Jul 15 01:50:42 2011 +0200
files:       binr/Makefile doc/maemo libr/Makefile maemo/Makefile maemo/README maemo/deb_hand.mak maemo/description.txt
description:
* Added support for new platform Maemo6-Harmattan
  - Linux/arm with debugger support
  - Debian package generation tools (no need for dpkg)
  - Compilation instructions at doc/maemo
* Fix install in binr and libr


changeset:   1618:dd1c50f121bf
user:        pancake@pair
date:        Thu Jul 14 22:14:11 2011 +0200
files:       configure configure.acr
description:
* Handle -lssl and pkg-config openssl
  - Should fix NetBSD build


changeset:   1617:4c6e8c50e7cf
user:        earada
date:        Thu Jul 14 16:17:39 2011 +0200
files:       libr/debug/p/debug_native.c libr/util/sys.c
description:
* Fix FreeBSD build


changeset:   1616:fd36edc76e42
user:        pancake
date:        Wed Jul 13 19:37:28 2011 +0200
files:       test/t/asm
description:
* Fix asm test


changeset:   1615:1ad69f14f1db
user:        pancake
date:        Wed Jul 13 19:28:53 2011 +0200
files:       test/tests.sh
description:
* Fix test suite


changeset:   1614:6c552be332cb
user:        pancake
date:        Wed Jul 13 19:22:38 2011 +0200
files:       binr/radare2/Makefile binr/radiff2/Makefile binr/rafind2/Makefile config-user.mk.acr configure configure.acr libr/debug/t/Makefile libr/hash/t/Makefile libr/include/r_userconf.h.acr libr/io/p/debug.mk libr/io/p/gdb.mk libr/io/p/haret.mk libr/io/t/Makefile libr/socket/Makefile libr/socket/t/Makefile
description:
* Use OpenSSL (cflags/ldflags) from pkg-config


changeset:   1613:8ecc34f29bc4
user:        pancake
date:        Wed Jul 13 18:55:16 2011 +0200
files:       doc/windows libr/asm/arch/msil/demsil.c libr/asm/p/asm_dalvik.c libr/bin/format/mach0/dyldcache.c libr/config.mk.tail libr/core/vmenus.c libr/include/list.h libr/include/r_types_base.h libr/io/p/io_w32dbg.c libr/socket/t/test.c libr/util/cache.c
description:
* Fix build and lot of warnings for w64


changeset:   1612:1552f58ae98c
user:        earada
date:        Wed Jul 13 17:41:26 2011 +0200
files:       binr/radare2/radare2.c libr/core/cmd.c libr/core/config.c libr/core/file.c libr/include/r_core.h
description:
* Use native api to calculate md5/sha1 hashes
- Fix bug 29


changeset:   1611:3dd0b00836fe
user:        pancake
date:        Wed Jul 13 13:59:55 2011 +0200
files:       binr/rax2/rax2.c doc/fortunes libr/fs/Makefile libr/fs/p/grub/Makefile libr/include/r_io.h libr/io/p/io_w32dbg.c
description:
* Fix w32 and w64 build (r_fs/grub issue)
* Dont quit rax2 when reading from stdin until 'q' or ^D
* Minor fix in fortunes


changeset:   1610:2637b35a1b31
user:        pancake
date:        Wed Jul 13 12:21:56 2011 +0200
files:       libr/core/cmd.c
description:
* Fix double pdf (thanks vext01)


changeset:   1609:5b507dbd2c5e
user:        pancake
date:        Wed Jul 13 11:30:25 2011 +0200
files:       Makefile
description:
* Remove manpages in purge
* Fix make symstall


changeset:   1608:980201d7e995
user:        pancake@pair
date:        Wed Jul 13 10:14:30 2011 +0200
files:       binr/binr.mk binr/radare2/Makefile binr/radiff2/Makefile binr/rafind2/Makefile doc/static libr/debug/p/libgdbwrap/gdbwrapper.c libr/debug/p/libgdbwrap/include/gdbwrapper-internals.h libr/debug/p/libgdbwrap/include/gdbwrapper.h libr/debug/t/Makefile libr/hash/t/Makefile libr/io/Makefile libr/io/p/debug.mk libr/io/p/gdb.mk libr/io/p/haret.mk libr/io/p/io_gdb.c libr/io/p/mach.mk libr/io/p/malloc.mk libr/io/p/shm.mk libr/io/t/Makefile libr/socket/t/Makefile mk/gcc.mk
description:
* Fix static build (--without-pic --with-nonpic)


changeset:   1607:3a06619b5226
user:        pancake@pair
date:        Tue Jul 12 23:22:33 2011 +0200
files:       libr/core/asm.c libr/core/cmd.c
description:
* Fix pi and /c commands


changeset:   1606:e220fccf6bc4
user:        earada
date:        Tue Jul 12 23:06:06 2011 +0200
files:       libr/core/cmd.c
description:
* Fix some warnings


changeset:   1605:d15cbb5291a3
user:        pancake
date:        Tue Jul 12 19:42:20 2011 +0200
files:       doc/windows libr/io/p/io_w32dbg.c mk/x86_64-w64-mingw32-gcc.mk
description:
* Fix build for ming-w64


changeset:   1604:026708c1cb88
user:        earada
date:        Tue Jul 12 17:26:56 2011 +0200
files:       libr/core/vmenus.c libr/print/format.c
description:
* Enhacements in keybindings used in visual mounts menus
* Implement missing 'pf p' pointer format


changeset:   1603:71f82c98175a
user:        pancake@pairch

binr/ranal2/ranal2.c
#include <r_types.h>
#include <r_anal.h>
#include <r_lib.h>
#include <r_util.h>
#include <getopt.h>

/* anal callback */
static int __lib_anal_cb(struct r_lib_plugin_t *pl, void *user, void *data) {
	RAnal *anal = (RAnal *)user;
	RAnalPlugin *plugin = (RAnalPlugin *)data;
	r_anal_add (anal, plugin);
	return R_TRUE;
}

static int __lib_anal_dt(struct r_lib_plugin_t *pl, void *p, void *u) {
	return R_TRUE;
}

static char *stackop2str(int type) {
	switch (type) {
		case R_ANAL_STACK_NULL:     return strdup ("null");
		case R_ANAL_STACK_NOP:      return strdup ("nop");
		case R_ANAL_STACK_INCSTACK: return strdup ("incstack");
		case R_ANAL_STACK_GET:      return strdup ("get");
		case R_ANAL_STACK_SET:      return strdup ("set");
	}
	return strdup ("unknown");
}

static char *optype2str(int type) {
	switch (type) {
		case R_ANAL_OP_TYPE_JMP:   return strdup ("jmp");
		case R_ANAL_OP_TYPE_UJMP:  return strdup ("ujmp");
		case R_ANAL_OP_TYPE_CJMP:  return strdup ("cjmp");
		case R_ANAL_OP_TYPE_CALL:  return strdup ("call");
		case R_ANAL_OP_TYPE_UCALL: return strdup ("ucall");
		case R_ANAL_OP_TYPE_REP:   return strdup ("rep");
		case R_ANAL_OP_TYPE_RET:   return strdup ("ret");
		case R_ANAL_OP_TYPE_ILL:   return strdup ("ill");
		case R_ANAL_OP_TYPE_NOP:   return strdup ("nop");
		case R_ANAL_OP_TYPE_MOV:   return strdup ("mov");
		case R_ANAL_OP_TYPE_TRAP:  return strdup ("trap");
		case R_ANAL_OP_TYPE_SWI:   return strdup ("swi");
		case R_ANAL_OP_TYPE_UPUSH: return strdup ("upush");
		case R_ANAL_OP_TYPE_PUSH:  return strdup ("push");
		case R_ANAL_OP_TYPE_POP:   return strdup ("pop");
		case R_ANAL_OP_TYPE_CMP:   return strdup ("cmp");
		case R_ANAL_OP_TYPE_ADD:   return strdup ("add");
		case R_ANAL_OP_TYPE_SUB:   return strdup ("sub");
		case R_ANAL_OP_TYPE_MUL:   return strdup ("mul");
		case R_ANAL_OP_TYPE_DIV:   return strdup ("div");
		case R_ANAL_OP_TYPE_SHR:   return strdup ("shr");
		case R_ANAL_OP_TYPE_SHL:   return strdup ("shl");
		case R_ANAL_OP_TYPE_OR:    return strdup ("or");
		case R_ANAL_OP_TYPE_AND:   return strdup ("and");
		case R_ANAL_OP_TYPE_XOR:   return strdup ("xor");
		case R_ANAL_OP_TYPE_NOT:   return strdup ("not");
		case R_ANAL_OP_TYPE_STORE: return strdup ("store");
		case R_ANAL_OP_TYPE_LOAD:  return strdup ("load");
	}
	return strdup ("unknown");

}

static int analyze(RAnal *anal, RAnalOp *op, ut64 offset, ut8* buf, int len) {
	char *bytes, *optype = NULL, *stackop = NULL;
	int ret;

	ret = r_anal_op (anal, op, offset, buf, len);
	if (ret) {
		stackop = stackop2str (op->stackop);
		optype = optype2str (op->type);
		bytes = r_hex_bin2strdup (buf, ret);
		printf ("bytes:    %s\n", bytes);
		printf ("type:     %s\n", optype);
		if (op->jump != -1LL)
			printf ("jump:     0x%08"PFMT64x"\n", op->jump);
		if (op->fail != -1LL)
			printf ("fail:     0x%08"PFMT64x"\n", op->fail);
		if (op->ref != -1LL)
			printf ("ref:      0x%08"PFMT64x"\n", op->ref);
		if (op->value != -1LL)
			printf ("value:    0x%08"PFMT64x"\n", op->value);
		printf ("stackop:  %s\n", stackop);
		printf ("stackptr: %"PFMT64d"\n", op->stackptr);
		printf ("decode str: %s\n", r_anal_op_to_string (anal, op));
		printf ("--\n");
		free (optype);
		free (stackop);
		free (bytes);
	}
	return ret;
}

static int usage() {
	printf ("ranal2 [opts] hexpairs|-\n"
			" -a [arch]    Set architecture plugin\n"
			" -b [bits]    Set architecture bits\n"
			" -B           Binary input (-l is mandatory for binary input)\n"
			" -h           This help\n"
			" -l [len]     Input length\n"
			" -L           List supported analysis plugins\n"
			" -o [offset]  Offset where this opcode is suposed to be\n"
			" If the last argument is '-' reads from stdin\n");
	return 1;
}

int main(int argc, char **argv) {
	RLib *lib;
	RAnal *anal = r_anal_new ();
	RAnalOp *op = r_anal_op_new ();
	ut8 *ptr, *buf = NULL, *data = NULL;
	ut64 offset = 0x8048000LL;
	char *arch = NULL;
	int bin = R_FALSE, len = 0, bits = 32;
	int c, idx, ret, tlen, word;

	lib = r_lib_new ("radare_plugin");
	r_lib_add_handler (lib, R_LIB_TYPE_ANAL, "analysis plugins",
		&__lib_anal_cb, &__lib_anal_dt, anal);
	r_lib_opendir (lib, r_sys_getenv ("LIBR_PLUGINS"));

	while ((c = getopt (argc, argv, "a:b:Bhl:Lo:")) != -1) {
		switch (c) {
		case 'a':
			arch = optarg;
			break;
		case 'b':
			bits = r_num_math (NULL, optarg);
			break;
		case 'B':
			bin = R_TRUE;
			break;
		case 'h':
			return usage ();
		case 'l':
			len = r_num_math (NULL, optarg);
			break;
		case 'L':
			return r_anal_list (anal);
		case 'o':
			offset = r_num_math (NULL, optarg);
			break;
		}
	}
	if (!argv[optind] || (bin && !len))
		return usage ();
	/* Set default options */
	if (arch) {
		if (!r_anal_use (anal, arch)) {
			eprintf ("Invalid plugin\n");
			return 1;
		}
	} else r_anal_use (anal, "x86");
	if (!r_anal_set_bits (anal, bits))
		r_anal_set_bits (anal, 32);
	/* Get input & convert to bin if necessary */
	if (argv[optind][0] == '-') {
		idx = 0;
		while (R_TRUE) {
			if (!(buf = realloc (buf, idx+1024)))
				return 1;
			fgets ((char*)buf+idx, 1024, stdin);
			if ((!bin && feof (stdin)) ||(len && idx >= len))
				break;
			idx += 1023;
		}
	} else {
		if (!(buf = (ut8 *)strdup (argv[optind])))
			return 1;
	}
	if (bin)
		data = (ut8*)buf;
	else {
		ptr = buf, word = tlen = 0;
		while (ptr[0]) {
			if (ptr[0]!= ' ' && ptr[0]!= '\n' && ptr[0]!= '\r')
				if (0==(++word%2))tlen++;
			ptr += 1;
		}
		data = malloc (tlen);
		if (!data) {
			r_anal_free (anal);
			r_anal_op_free (op);
			return 1;
		}
		r_hex_str2bin ((char *)buf, data);
		if (!len || len > tlen) len = tlen;
		free (buf);
	}
	/* Analyze */
	for (idx=ret=0; idx<len; idx+=ret) {
		if (!(ret = analyze (anal, op, offset+idx, data+idx, len-idx))) {
			eprintf ("Ooops\n");
			free (data);
			r_anal_free (anal);
			r_anal_op_free (op);
			return 1;
		}
	}
	free (data);
	r_anal_free (anal);
	r_anal_op_free (op);
	return 0;
}
